cpu/atmega_common/periph/uart.c

```diff
@@ -123,44 +143,122 @@ int uart_init(uart_t uart, uint32_t baudrate, uart_rx_cb_t rx_cb, void *arg)
     isr_ctx[uart].rx_cb = rx_cb;
     isr_ctx[uart].arg   = arg;
 
+    /* TODO: What about just UART (without any number?) */
+    #ifdef MEGA_UART0
+    if(ctx[uart].dev == MEGA_UART0) {
+        ctx[uart].num = 0;
+    } else
+    #endif
+    #ifdef MEGA_UART1
+    if(ctx[uart].dev == MEGA_UART1) {
+        ctx[uart].num = 1;
+    } else
+    #endif
+    #ifdef MEGA_UART2
+    if(ctx[uart].dev == MEGA_UART3) {
+        ctx[uart].num = 2;
+    } else
+    #endif
+    #ifdef MEGA_UART3
+    if(ctx[uart].dev == MEGA_UART3) {
+        ctx[uart].num = 3;
+    } else
+    #endif
+
+    if (ctx[uart].num == -1 ) {
+        return -1;
+    }
+
     /* disable and reset UART */
-    dev[uart]->CSRB = 0;
-    dev[uart]->CSRA = 0;
+    ctx[uart].dev->CSRB = 0;
+    ctx[uart].dev->CSRA = 0;
 
     /* configure UART to 8N1 mode */
-    dev[uart]->CSRC = (1 << UCSZ00) | (1 << UCSZ01);
+    ctx[uart].dev->CSRC = (1 << UCSZ00) | (1 << UCSZ01);
+    /* Power on uart so we can use it */
+    uart_poweron(uart);
     /* set clock divider */
     _set_brr(uart, baudrate);
 
-    /* enable RX and TX and the RX interrupt */
-    if (rx_cb) {
-        dev[uart]->CSRB = ((1 << RXCIE0) | (1 << RXEN0) | (1 << TXEN0));
-    }
-    else {
-        dev[uart]->CSRB = (1 << TXEN0);
-    }
-
-
     return UART_OK;
 }
 
 void uart_write(uart_t uart, const uint8_t *data, size_t len)
 {
     for (size_t i = 0; i < len; i++) {
-        while (!(dev[uart]->CSRA & (1 << UDRE0))) {};
-        dev[uart]->DR = data[i];
+        while (!(ctx[uart].dev->CSRA & (1 << UDRE0))) {};
+        ctx[uart].dev->DR = data[i];
     }
 }
 
 static inline void isr_handler(int num)
 {
-    isr_ctx[num].rx_cb(isr_ctx[num].arg, dev[num]->DR);
+    isr_ctx[num].rx_cb(isr_ctx[num].arg, ctx[num].dev->DR);
 
     if (sched_context_switch_request) {
         thread_yield();
     }
 }
 
+void uart_poweron(uart_t uart) {
+    switch(ctx[uart].num) {
+#ifdef MEGA_UART0
```

Comment:
Do we need a differentiation if we have rx/tx enabled or not? See discussion in #7947